src/hbbft.erl

```diff
@@ -202,52 +202,81 @@ handle_msg(Data = #hbbft_data{round=R}, J, {{acs, R}, ACSMsg}) ->
                                         SerializedShare = hbbft_utils:share_to_binary(Share),
                                         {multicast, {dec, Data#hbbft_data.round, I, SerializedShare}}
                                 end, Results),
-            {Data#hbbft_data{acs=NewACS, acs_results=Results}, {send,  hbbft_utils:wrap({acs, Data#hbbft_data.round}, ACSResponse) ++ Replies}};
+            %% verify any shares we received before we got the ACS result
+            VerifiedShares = maps:map(fun({I, _}, {undefined, Share}) ->
+                                              case lists:keyfind(I, 1, Results) of
+                                                  {I, Enc} ->
+                                                      EncKey = get_encrypted_key(Data#hbbft_data.secret_key, Enc),
+                                                      Valid = tpke_pubkey:verify_share(tpke_privkey:public_key(Data#hbbft_data.secret_key), Share, EncKey),
+                                                      {Valid, Share};
+                                                  false ->
+                                                      %% this is a share for an RBC we will never decode
+                                                      {undefined, Share}
+                                              end;
+                                         (_, V) ->
+                                              V
+                                      end, Data#hbbft_data.dec_shares),
+            {Data#hbbft_data{acs=NewACS, acs_results=Results, dec_shares=VerifiedShares}, {send,  hbbft_utils:wrap({acs, Data#hbbft_data.round}, ACSResponse) ++ Replies}};
         {NewACS, defer} ->
             {Data#hbbft_data{acs=NewACS}, defer}
     end;
 handle_msg(Data = #hbbft_data{round=R}, _J, {dec, R2, _I, _Share}) when R2 > R ->
     {Data, defer};
 handle_msg(Data = #hbbft_data{round=R}, J, {dec, R, I, Share}) ->
-    %% the Share now is a binary, deserialize it and then store in the dec_shares map
-    DeserializedShare = hbbft_utils:binary_to_share(Share, Data#hbbft_data.secret_key),
-    NewShares = maps:put({I, J}, DeserializedShare, Data#hbbft_data.dec_shares),
     %% check if we have enough to decode the bundle
-    SharesForThisBundle = [ S || {{Idx, _}, S} <- maps:to_list(NewShares), I == Idx],
-    case length(SharesForThisBundle) > Data#hbbft_data.f andalso not maps:is_key({I, J}, Data#hbbft_data.dec_shares) andalso lists:keymember(I, 1, Data#hbbft_data.acs_results) of
+    case maps:is_key({I, J}, Data#hbbft_data.dec_shares) orelse maps:is_key(I, Data#hbbft_data.decrypted) of
```

Comment:
I'm just wondering if it would be better to separate these out or change the order of conditions to fail faster? If you already have decrypted no need to check dec_shares presumably?